### PR TITLE
Fixed issue #18337: Placeholders VALIDFROM and VALIDUNTIL at email templates do not use localized dates

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Comment by Shnoulle** [comment]
1. It's a new feature
2. Unsure we need it : because some admin user can test year for example and show something.
3. date_format(VALIDFROM) and date_format(VALIDUNTIL) can be used (if it work)

**Comment by gabrieljenik** [comment]
> It's a new feature

@maziminke thoughts?

**Comment by Shnoulle** [comment]
Else  : it can broke some current usage of VALIDFROM : 

- comparaison (VALIDUNTIL < EXPIRY) for example
- `{date("d/m/Y",strtotime(VALIDFROM))}` usage

We don't update EXPIRY, allow user to use `{date("d/m/Y",strtotime(EXPIRY))}` then ;: i think we MUST not update it …

**Comment by gabrieljenik** [comment]
> PS : if you send it according to token language : user can not use their own format : you remove a solution

Yes, you are right.
Still, now admins can't show dates in the matching language.
If doing a survey for US and Spain, that could lead to many confusions to participants as they use opposte formats: `d/m/y` vs `m/d/y`. Taking into account that, could be considered a bug.

@maziminke I believe that either we introduce new replacement variables formatted according to the current language or survey admins use something like `{date("d/mmm/Y",strtotime(VALIDFROM))}` which could be international.

If you need a quick solution, I suggest the latter.

**Comment by Shnoulle** [comment]
> If doing a survey for US and Spain

`{if(TOKEN;LANGUAGE=="en",date("d/m/Y",strtotime(VALIDFROM)),date("m/d/Y",strtotime(VALIDFROM)))}`

> introduce new replacement variables formatted according to the current language

Yes : like VALIDFROMFORMATTED, unusure really needed.

> introduce new function that format dates according to the current language

You mean add a expression manager function format_date(); ?
Must define the option 

```
@param string $date ISO format
@param booelan $withTime : default to ? false
@param language $language : default to survey or token language
```

A [good core plugin](https://manual.limesurvey.org/ExpressionManagerStart#Function_definition) to do here :+1:

**Comment by gabrieljenik** [comment]
> > If doing a survey for US and Spain
> 
> `{if(TOKEN;LANGUAGE=="en",date("d/m/Y",strtotime(VALIDFROM)),date("m/d/Y",strtotime(VALIDFROM)))}`

Sure. I was providing an example on why I think this is a bug.
Also, on the other side, f I have a 20 language survey, what should I do?

**Comment by Shnoulle** [comment]
> > > If doing a survey for US and Spain
> > 
> > 
> > `{if(TOKEN;LANGUAGE=="en",date("d/m/Y",strtotime(VALIDFROM)),date("m/d/Y",strtotime(VALIDFROM)))}`
> 
> Sure. I was providing an example on why I think this is a bug. Also, on the other side, f I have a 20 language survey, what should I do?

VALIDFROMFORMATTED : **but** still the issue with time or not … 

or
`
{if(TOKEN;LANGUAGE=="en",date("d/m/Y",strtotime(VALIDFROM)))}{if(TOKEN;LANGUAGE=="en",date("m/d/Y",strtotime(VALIDFROM)))}…` easy to do with Calc ;)

**Comment by Shnoulle** [comment]
> > > If doing a survey for US and Spain
> > 
> > 
> > `{if(TOKEN;LANGUAGE=="en",date("d/m/Y",strtotime(VALIDFROM)),date("m/d/Y",strtotime(VALIDFROM)))}`
> 
> Sure. I was providing an example on why I think this is a bug. Also, on the other side, f I have a 20 language survey, what should I do?

Else with new expression manager function `{format_date(VALIDFROM,0,TOKEN:LANGUAGE)}`

File: application/core/plugins/dateFunctions/EMFunctions.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+class EMFunctions
+{
+    /**
+     * Formats a date according to the Survey's date format for the specified language
+     * @param string $date : a date in "Y-m-d H:i:s" format. Example: VALIDFROM.
+     * @param string $language : the language used for localization. If the language is not configured in the survey, the base language will be used. Example: TOKEN:LANGUAGE
+     * @return string
+     */
+    public static function localize_date($date, $language)
```
Comment (Shnoulle) [inline application/core/plugins/dateFunctions/EMFunctions.php:21]: In my opinion : 

1. $language must be optional : get it for App()->getLanguage() if null
2. add an opt for time

**Review by Shnoulle — COMMENTED** [review]
App()->getLanguage() by default and need an option for time.

File: application/core/plugins/dateFunctions/EMFunctions.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $surveyId = LimeExpressionManager::getLEMsurveyId();
+        $survey = Survey::model()->findByPk($surveyId);
+
+        if (empty($survey)) {
+            return '';
+        }
+
+        // If the specified language is note one of the survey's languages, fallback to the survey's base language.
+        if (!in_array($language, $survey->getAllLanguages()) || empty($survey->languagesettings[$language])) {
+            $language = $survey->language;
```
Comment (Shnoulle) [inline application/core/plugins/dateFunctions/EMFunctions.php:36]: Then allow null language ;)

File: application/core/plugins/dateFunctions/EMFunctions.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+     * @return string
+     */
+    public static function localize_date($date, $language = null)
+    {
+        if (empty($date)) {
+            return '';
+        }
+
+        // Try to get Survey ID from the EM
+        $surveyId = LimeExpressionManager::getLEMsurveyId();
```
Comment (gabrieljenik) [inline application/core/plugins/dateFunctions/EMFunctions.php:28]: when called from email, this variable is not set on EM. That;s why we keep on trying for other alternatives if no SID found.

Comment (Shnoulle) [reply application/core/plugins/dateFunctions/EMFunctions.php:28]: Maybe we can create a Dev mantis issue for this.

It set in `\Yii::app()->session['LEMsid'] ` then seems OK for me.

**Review by Shnoulle — APPROVED** [review]
Adding time by option can be done after.

**Comment by Shnoulle** [comment]
@maziminke you have some time to test ?

**Comment by maziminke** [comment]
Sorry, no time for testing currently :-(

**Comment by gabrieljenik** [comment]
This is tested. When using the function on an email, the language need to be explicitly set as parameter, either hard coded or using a replacement variable.

Ex:
> You can start from {localize_date(TOKEN:VALIDFROM, LANGUAGE)} and you have until {localize_date(TOKEN:VALIDUNTIL, TOKEN:LANGUAGE)} to take the survey.

**Comment by maziminke** [comment]
We should definitely add a note at the manual about the new feature and how to use it.
Can you add this? I can update the German version afterwards.

**Comment by Shnoulle** [comment]
Don't update manual if it's not merged

**Comment by gabrieljenik** [comment]
> We should definitely add a note at the manual about the new feature and how to use it. Can you add this? I can update the German version afterwards.

This is a Core plugin. Where should we add it on the manual?

**Comment by Shnoulle** [comment]
> > We should definitely add a note at the manual about the new feature and how to use it. Can you add this? I can update the German version afterwards.
> 
> This is a Core plugin. Where should we add it on the manual?

https://manual.limesurvey.org/Category:Core_Plugins

**Comment by gabrieljenik** [comment]
> We should definitely add a note at the manual about the new feature and how to use it.

https://manual.limesurvey.org/DateFunctions

**Comment by olleharstedt** [comment]
New feature should go to dev branch?

**Comment by Shnoulle** [comment]
Oh yes :)
Not a lot of commit on current dev, sorry :)

**Comment by gabrieljenik** [comment]
> New feature should go to dev branch?

You are right. Still, as it is wrapped in a plugin and solves a current bug, I thought it could be OK to merge into master.
Let me know if I should change that

**Pull request merged**